kvcoord: flush write buffer on unbufferable write

stevendanna opened the pull request:
To avoid bugs such as #149911 we want to flush the buffer if we might end up with a pipelined write. Currently, this is only true when we are configured to skip transforming locking scans.

This has the nice property that a given transaction either has buffered writes or has pipelined write, but never both.

Fixes #149911

Epic: none

stevendanna committed:
kvcoord: flush write buffer on unbufferable write

To avoid bugs such as #149911 we want to flush the buffer if we might
end up with a pipelined write. Currently, this is only true when we
are configured to skip transforming locking scans.

This has the nice property that a given transaction either has
buffered writes or has pipelined write, but never both.

Fixes #149911

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149927)
<!-- Reviewable:end -->

stevendanna committed:
kvcoord: make kv.transaction.write_buffering.transformations.scans.enabled metamorphic

Epic: none
Release note: None

stevendanna committed:
kvcoord: make kv.transaction.write_buffering.max_buffer_size metamorphic

Epic: none
Release note: None

stevendanna committed:
kvcoord: rename a few tests

This allows a developer to write a test filter that calls all of the
txn write buffer unit tests.

Epic: none
Release note: None

stevendanna commented:
The other problem is fixed here: https://github.com/cockroachdb/cockroach/pull/150276

stevendanna committed:
kvcoord: add a couple of missing test cases

Epic: none
Release note: None

stevendanna commented:
@arulajmani @miraradeva Even if we do the "writeBuffer xor pipelining" patch, this still seems reasonable to me for master: At worst, it only affects a few users who turn off scan transforms. Also, it might be preferable to disabling pipelining completely because keeping pipelining enabled allows for a more graceful fallback to pipelining if a DeleteRange is encountered. This is a long way to say, this is probably still worth reviewing.

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just a few requests for comments.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149927#-:-OVIS2jBDvGhgZJV7qlC:b-lopxu5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149927#-OVIOb3wEd0dv7Z9eH7L:-OVIOb3wEd0dv7Z9eH7M:b4at60e):*
Maybe one more sentence here: if a locking scan with replicated durability is transformed, the replicated locks are transformed into unreplicated locks; otherwise, the replicated locks are pipelined.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149927#-OVIPLeB2LMm4KXFEHHO:-OVIPLeB2LMm4KXFEHHP:b-cd87rk) ([raw file](https://github.com/cockroachdb/cockroach/blob/842117d0fb9139f9d4e632a38b228bcf0da2ff23/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L317)):*
> ```Go
> 	}
> 
> 	assertNoWrites(transformedBa)
> ```

nit: A comment to explain why we don't expect any writes past this point? And whether replicated locks count as writes.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 340 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149927#-OVIPgtg47kkr_9VzY86:-OVIPgtg47kkr_9VzY87:b3nrb9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/842117d0fb9139f9d4e632a38b228bcf0da2ff23/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L340)):*
> ```Go
> 			//
> 			// We've opted here to flush our batch whenever we may issue a batch that
> 			// results in a pipelined write.
> ```

nit: Not just any pipelined write, right? Specifically, pipelined locking scan. Maybe also mention that this is not an issue for replicated locking Gets because we always transform them?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 491 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149927#-OVIQTnS7YfCz-A1T5V1:-OVIQTnS7YfCz-A1T5V2:bduk4qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/842117d0fb9139f9d4e632a38b228bcf0da2ff23/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L491)):*
> ```Go
> 			estimate += lockKeyInfoSize
> 		case *kvpb.GetRequest:
> 			if IsReplicatedLockingRequest(t) {
> ```

We weren't checking `KeyLockingStrength > lock.None` before here; was this a bug?


<!-- Sent from Reviewable.io -->

stevendanna commented:
> I assume it's too hard to repro the full kvnemesis failure with the in-flight writes manipulation and the lost write?

To reproduce the bug, you need to pipeline the write and then arrange for that write to be lost. But here, we no longer pipeline the write at all so it sort of invalidates the test.

> Maybe it's possible to do a simpler unit test for maybeStripInFlightWrites to highlight that it can't handle duplicate seqnums?

Maybe also a test-only assertion that checks for writes at the same sequence number in InFlightWrites somewhere?

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149927)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 340 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149927#-OVIPgtg47kkr_9VzY86:-OVIVk7s2xOx56xLXrFf:b2t1v9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/842117d0fb9139f9d4e632a38b228bcf0da2ff23/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L340)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

nit: Not just any pipelined write, right? Specifically, pipelined locking scan. Maybe also mention that this is not an issue for replicated locking Gets because we always transform them?
</blockquote></details>

I _think_ the net result of this change is that now we flush if we encounter any write that can't be buffered, which implies we flush if we hit any pipeline-able write that can't be buffered since they are a subset of writes. The new assertion confirms this in test builds.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 491 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149927#-OVIQTnS7YfCz-A1T5V1:-OVIWAqU4_nPQ--4OeJM:b-4s05ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/842117d0fb9139f9d4e632a38b228bcf0da2ff23/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L491)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

We weren't checking `KeyLockingStrength > lock.None` before here; was this a bug?
</blockquote></details>

_Maybe_ for hand-constructed batches, but I think that SQL ends up only setting a non-zero Durability if lock strength is > lock.None so for SQL the check is redundant.


<!-- Sent from Reviewable.io -->
